### PR TITLE
Add body param extraction for array of hashes

### PR DESCRIPTION
**Pull request by bautrey37**
This change is specifically to address extracting the webhook param in
engagement-api. Webhooks are specified as an array of hashes. The
previous functionality would lead to the array of hashes to have strings
for keys, while the `webhooks` key is a symbol.

The new version of dry-schema wants all keys to be a symbol. So when a
key was a string, the key was not found and the validation failed.

This implementation checks if the key's result is an array, to symbolize
all keys in the nested array of hashes structure. There should be no
need to specify all the keys in the webhooks as these are validated with
dry-schema.

MSG-75
MSG-71

### CODE REVIEW

File: spec/salestation/web/extractors/body_param_extractor_spec.rb [hunk 1]
```diff
@@ -102,4 +102,18 @@ def initialize(env_data)
...
+    let(:params) { {'x' => 'y', 'webhooks' => [ {'foo' => 'bar'} ] } }
+    let(:expected_result) { {x: 'y', webhooks: [ {foo: 'bar'} ] } }
+
+    it 'extracts body params from request' do
+      result = extract_body_params
+
+      expect(result).to be_a(Deterministic::Result::Success)
+      expect(result.value).to eq(expected_result)
+    end
+  end
```
Comment (bautrey37) [inline spec/salestation/web/extractors/body_param_extractor_spec.rb:118]: I'll be adding more tests to test for negative scenarios.

File: lib/salestation/web/extractors.rb [hunk 2]
```diff
@@ -219,14 +222,19 @@ def self.[](*filters)
       # Extracts and symbolizes params from request body
       #
       # @example
-      #  extractor = BodyParamExtractor[:x, :y, {foo: [:bar, :baz]}]
+      #  extractor = BodyParamExtractor[:x, :y, {foo: [:bar, :baz]}, aaa: ?? ]
```
[the remark this answers is not in the document]
Comment (bautrey37) [reply lib/salestation/web/extractors.rb]: Fixed

**Comment by bautrey37** [comment]
Usage of changes in https://github.com/salemove/engagement-registry/pull/1536

File: lib/salestation/web/extractors.rb [hunk 3]
```diff
@@ -197,12 +197,28 @@ def self.extract(filters, request_hash)
             when Symbol
               stringified_key = filter.to_s
               extracted_data[filter] = request_hash[stringified_key] if request_hash.key?(stringified_key)
+              if extracted_data[filter].is_a?(Array)
```
Comment (SteamRiC) [inline lib/salestation/web/extractors.rb:200]: would calling
```
input = extract(filters, request_hash).deep_symbolize_keys
```
on line 189 also work instead of manually symbolizing the keys here?

**Comment by SteamRiC** [comment]
i think we should also make it backwards compatible as this can break other services that fetch the latest salestation right?

File: lib/salestation/web/extractors.rb (hunk 3 above)
Comment (bautrey37) [reply lib/salestation/web/extractors.rb:200]: `deep_symbolize_keys` is a rails function. This library doesn't have access to rails as I know.

Comment (SteamRiC) [reply lib/salestation/web/extractors.rb:200]: we get it from `activesupport` library usually, but it's indeed not included here. so nvm

File: lib/salestation/web/extractors.rb [hunk 4]
```diff
@@ -197,12 +197,28 @@ def self.extract(filters, request_hash)
...
             when Hash
               filter.each do |key, nested_filters|
                 stringified_key = key.to_s
                 if request_hash.key?(stringified_key)
                   value = request_hash.fetch(stringified_key)
-                  extracted_data[key] = value.nil? ? nil : extract(nested_filters, value)
+                  extracted_data[key] =
+                    if !value.is_a?(Hash)
+                      value
+                    elsif value.nil?
```
[the remark this answers is not in the document]
Comment (bautrey37) [reply lib/salestation/web/extractors.rb]: Yep, your right. https://github.com/salemove/salestation/pull/62/commits/355b6b79cf63f813263ea2669aba4e1be7c5c875

**Comment by bautrey37** [comment]
It can break other services who take the latest salestation in that the keys that used to be a string, are now symbols. This usually breaks when something is done like `input.fetch('key')`. The Schema.Params works with both string and symbol keys. To have a mix of string and symbol keys is really confusing and sticking to one type is preferred. So I think its good to enforce a standard than to make everything backward compatible. 

If you want to make it backwards compatible, how would that be done?

**Review by SteamRiC — APPROVED** [review]
